CI: Update Used Actions' Versions

BowTiedDevOps opened the pull request:
## Description
Updated all marketplace actions to their latest version - the node version that some of them were using was `Node.js 16`, which is deprecated and comes up as a warning in the runners, `Node.js 20` being the one used in the latest versions of these actions.

**Notes:**
- `upload-artifact` action had a change from `v3` to `v4` and the workflow is unable to append to the initial artifact when uploading with the same name, and I've made the composite action upload artifacts with different names. Because of this, the `download-artifact` action had to be updated to only download artifacts with a specific pattern (`<tag>-binary-build-*` in the case of release workflow).
- Added input for artifact pattern (`<tag>-binary-build-*`) to `generate-checksum` composite action, because previously it was downloading an artifact with hardcoded name (`artifact`) - which no longer exists (explained above).

Example run (the bitcoin test failure is not affected by the changes in this PR): https://github.com/BowTiedDevOps/stacks-core/actions/runs/9082502434

Release after updating the workflows: https://github.com/BowTiedDevOps/stacks-core/releases/tag/2.5.0.0.4-rc4

This PR is to be merged after the `actions` repository one: https://github.com/stacks-network/actions/pull/40

CharlieC3 reviewed (dismissed):
LGTM :shipit:

wileyj commented on .github/workflows/ci.yml:
```diff
@@ -50,11 +50,11 @@ jobs:
     steps:
       - name: Checkout the latest code
         id: git_checkout
-        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+        uses: actions/checkout@44c2b7a8a4ea60a981eaca3cf939b5f4305c123b # v4.1.5
 
       - name: Setup Rust Toolchain
         id: setup_rust_toolchain
-        uses: actions-rust-lang/setup-rust-toolchain@f3c84ee10bf5a86e7a5d607d487bf17d57670965 # v1.5.0
+        uses: actions-rust-lang/setup-rust-toolchain@b113a30d27a8e59c969077c0a0168cc13dab5ffc # v1.8.0
         with:
           components: rustfmt
           cache: false
```
this seems odd to keep here now that the rustfmt workflow was moved to the actions repo. 
if we moved these steps there, i think we can avoid having to change the ci workflow in the future. 
i'll create a new issue an assign

wileyj reviewed (dismissed):
shipit

BowTiedDevOps commented:
Moved the setting of the rust toolchain and checkout-ing the code inside composite actions for easier code maintenance, changed in https://github.com/stacks-network/actions/pull/41.
Closes #4794 

Example run (rustfmt): https://github.com/BowTiedDevOps/stacks-core/actions/runs/9087322966/job/24974854258

wileyj reviewed (approved):
shipit

wileyj commented:
fyi - rerunning workflows after the required change was merged in the actions repo

blockstack-devops commented:
This pull request has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
